internal/verifier/compare.go

```diff
@@ -25,31 +25,40 @@ func (verifier *Verifier) FetchAndCompareDocuments(
 	types.ByteCount,
 	error,
 ) {
-	srcChannel, dstChannel, readSrcCallback, readDstCallback := verifier.getFetcherChannelsAndCallbacks(task)
+	var srcChannel, dstChannel <-chan bson.Raw
+	var readSrcCallback, readDstCallback func(context.Context, *retry.FuncInfo) error
 
 	results := []VerificationResult{}
 	var docCount types.DocumentCount
 	var byteCount types.ByteCount
 
 	retryer := retry.New(retry.DefaultDurationLimit)
 
-	err := retryer.Run(
-		givenCtx,
-		verifier.logger,
-		readSrcCallback,
-		readDstCallback,
-		func(ctx context.Context, _ *retry.FuncInfo) error {
-			var err error
-			results, docCount, byteCount, err = verifier.compareDocsFromChannels(
-				ctx,
-				task,
-				srcChannel,
-				dstChannel,
-			)
+	err := retryer.
+		WithBefore(func() {
+			srcChannel, dstChannel, readSrcCallback, readDstCallback = verifier.getFetcherChannelsAndCallbacks(task)
+		}).
```

Comment:
I'm trying to understand why do we need before func. Is it the same as calling the function at the start of the retried function except that the error from it won't be retried?

Reply:
The trick is that there are _multiple_ retried funcs. Having a “before” func guarantees execution of one piece before all the others … which is crucial here because the callbacks & channels have to get recreated every time.

I’ll clarify this in docs.